crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -1,378 +1,503 @@
     }
 
-    fn find_all_successors(&self, block_id: BasicBlockId) -> HashSet<BasicBlockId> {
-        let mut stack = vec![];
-        let mut visited = HashSet::new();
-
-        // Fetch initial block successors
-        let successors = self.cfg.successors(block_id);
-        for successor in successors {
-            if !visited.contains(&successor) {
-                stack.push(successor);
+    fn handle_terminator(
+        &self,
+        function: &mut Function,
+        block: BasicBlockId,
+        references: &mut Block,
+        last_stores: &mut BTreeMap<ValueId, InstructionId>,
+    ) {
+        match function.dfg[block].unwrap_terminator() {
+            TerminatorInstruction::JmpIf { .. } => (), // Nothing to do
+            TerminatorInstruction::Jmp { destination, arguments, .. } => {
+                let destination_parameters = function.dfg[*destination].parameters();
+                assert_eq!(destination_parameters.len(), arguments.len());
+
+                // Add an alias for each reference parameter
+                for (parameter, argument) in destination_parameters.iter().zip(arguments) {
+                    if function.dfg.type_of_value(*parameter) == Type::Reference {
+                        let argument = function.dfg.resolve(*argument);
+
+                        if let Some(expression) = references.expressions.get(&argument) {
+                            if let Some(aliases) = references.aliases.get_mut(expression) {
+                                // The argument reference is possibly aliased by this block parameter
+                                aliases.insert(*parameter);
+                            }
+                        }
+                    }
+                }
+            }
+            TerminatorInstruction::Return { return_values } => {
+                // Removing all `last_stores` for each returned reference is more important here
+                // than setting them all to ReferenceValue::Unknown since no other block should
+                // have a block with a Return terminator as a predecessor anyway.
+                self.mark_all_unknown(return_values, function, references, last_stores);
             }
         }
+    }
+}
 
-        // Follow the CFG to fetch the remaining successors
-        while let Some(block) = stack.pop() {
-            visited.insert(block);
-            let successors = self.cfg.successors(block);
-            for successor in successors {
-                if !visited.contains(&successor) {
-                    stack.push(successor);
+impl Block {
+    /// If the given reference id points to a known value, return the value
+    fn get_known_value(&self, address: ValueId) -> Option<ValueId> {
+        if let Some(expression) = self.expressions.get(&address) {
+            if let Some(aliases) = self.aliases.get(expression) {
+                // We could allow multiple aliases if we check that the reference
+                // value in each is equal.
```

Comment:
Should we make an issue for this or is it already being worked on?

Reply:
I'll make an issue. It shouldn't be too hard to implement, I just thought it was a somewhat rare case where a reference is aliased to multiple others which all are known to store the same value. Perhaps this assumption was incorrect though.

Reply:
#2443